Adds a11y labels and focus

amygroshek opened the pull request:
Adds a11y labels and focus to notify screenreader user to use
button to access additional content and set focus to newly
displayed content after button is selected.

amygroshek committed:
Adds a11y labels and focus

Adds a11y labels and focus to notify screenreader user to use
button to access additional content and set focus to newly
displayed content after button is selected.

danielstorey commented on templates/stacklist.hbs:
```diff
@@ -14,9 +14,13 @@
+                        <button class="stacklist-next" aria-label="{{_globals._accessibility._ariaLabels.next}}" role="button">
+                    {{/if}}
+                            {{{_button.startText}}}
+                        </button>
```
This seems a bit bulky. I'd prefer something like
```
var ariaLabel = _globals._components._stacklist.ariaButtonLabel || _globals._accessibility._ariaLabels.next;
this.model.set({buttonAriaLabel: ariaLabel});
```
in the view, and then  `<button class="stacklist-next" aria-label="{{buttonAriaLabel}}" role="button">` in the template.

PS do we still need the `role="button"` attribute for a button element?!

Before changing this however, am I right in thinking that screen readers will prioritise reading the aria label over the button text? Or will both be read? If the former then I wouldn't want the actual text to be missed, especially since there is an option to change this for each item. Also, since you've now included a reference to the button in the instruction - would this be enough since there is only one button?

amygroshek replied:
You're right that the content of the button won't be read. What I'd really like to do is revise/append the `aria-label` with whatever the text of that specific button is going to be, since the start text can be different from the subsequent buttons, according to the component values. so that you're only using `_globals._accessibility._ariaLabels.next` as a fallback if the labels aren't dictated in components.json **and** `_globals._components._stacklist.ariaButtonLabel` is empty. I haven't yet looked into where that's being done in the js.

danielstorey commented on js/adapt-stacklist.js:
```diff
@@ -74,6 +74,7 @@ define(function(require) {
                 duration: this.TRANSITION_TIME,
                 complete: function() {
 			        $item.addClass("show");
+					$item.a11y_focus();
```
Could you please change the indentation to 4 spaces :)

Also, not so bothered, but this could potentially be simplified to `$item.addClass("show").a11y_focus();` thanks to jQuery's method chaining :)

danielstorey reviewed (changes requested):
Hi Amy, thanks so much for your contribution. Please would you attend to my comments re code changes below.

With regard to my comments on whether to include an aria label on the button I'd like to hear your thoughts. I'll defer to your expertise on accessibility though as I've also seen your comments in the community.

Thanks, Dan

amygroshek commented:
@danielstorey thanks for the PR. I'll look at this later today. The `role="button"` was already in the template before my changes...  I think that because it's actually a `<button>` we don't really need that role. Maybe a `type="button"`? I'll ask my client about it.

amygroshek committed:
fixup! Adds a11y labels and focus

amygroshek committed:
fixup! Adds a11y labels and focus

amygroshek commented:
@danielstorey I've moved the logic for the `aria-label` into the JS as requested. I'm really struggling with the indentation in adapt-stacklist.js. That file is a mixture of hard and soft tabs. My indents are normalized once I convert the entire file to hard tabs, but then additional lines appear in the diff as changed. 

My testing with VoiceOver shows that the item is identified as a button even without `role="button"` or `type="button"`. I'm checking with my client but my guess is we're fine without the role.

danielstorey commented:
Hi @amygroshek, this looks great. If you can convert the file all to soft tabs that would be great. Don't worry about the diff. If you append `?w=1` to the url it ignores changes to whitespace. However, I'm equally happy to tidy this up at a later stage.

Could you just change your single quotes to double to maintain consistency with the rest of the file - then I'll merge. Thanks!

amygroshek committed:
fixup! Adds a11y labels and focus

amygroshek commented:
@danielstorey Changed single quotes, converted tabs to spaces, and added the chaining you had requested earlier. Somehow that change got dropped while I was fiddling with the indents...

amygroshek committed:
fixup! Adds a11y labels and focus

amygroshek committed:
fixup! Adds a11y labels and focus

amygroshek committed:
fixup! Adds a11y labels and focus

danielstorey commented:
Awesome. Thanks for your work on this :-)
